core/src/avm2/optimize.rs

```diff
@@ -0,0 +1,893 @@
+use swf::avm2::types::Index;
+
+#[derive(Clone, Copy, Debug)]
+enum ValueType<'gc> {
+    // Either a class, or null.
+    Class(ClassObject<'gc>),
+    Int,
+    Uint,
+    Number,
+    Boolean,
+    Null,
+    Any,
+}
+
+#[derive(Clone, Debug)]
+struct Locals<'gc>(Vec<ValueType<'gc>>);
+
+impl<'gc> Locals<'gc> {
+    fn new(size: usize) -> Self {
+        Self(vec![ValueType::Any; size])
+    }
+
+    fn set_class_object(&mut self, index: usize, class: ClassObject<'gc>) {
+        self.0[index] = ValueType::Class(class);
+    }
+
+    fn set_class(&mut self, index: usize, class: GcCell<'gc, Class<'gc>>) {
+        // FIXME: Getting the ClassObject this way should be unnecessary
+        // after the ClassObject refactor
+        self.0[index] = class
+            .read()
+            .class_object()
+            .map(ValueType::Class)
+            .unwrap_or(ValueType::Any);
+    }
+
+    fn set_any(&mut self, index: usize) {
+        self.0[index] = ValueType::Any;
+    }
+
+    fn set(&mut self, index: usize, value: ValueType<'gc>) {
+        self.0[index] = value;
+    }
+
+    fn at(&self, index: usize) -> Option<ValueType<'gc>> {
+        self.0.get(index).copied()
+    }
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Stack<'gc>(Vec<ValueType<'gc>>);
+
+impl<'gc> Stack<'gc> {
+    fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    fn push_class_object(&mut self, class: ClassObject<'gc>) {
+        self.0.push(ValueType::Class(class));
+    }
+
+    fn push_class(&mut self, class: GcCell<'gc, Class<'gc>>) {
+        // FIXME: Getting the ClassObject this way should be unnecessary
+        // after the ClassObject refactor
+        self.0.push(
+            class
+                .read()
+                .class_object()
+                .map(ValueType::Class)
+                .unwrap_or(ValueType::Any),
+        );
+    }
+
+    fn push_int(&mut self) {
+        self.0.push(ValueType::Int);
+    }
+
+    fn push_uint(&mut self) {
+        self.0.push(ValueType::Uint);
+    }
+
+    fn push_number(&mut self) {
+        self.0.push(ValueType::Number);
+    }
+
+    fn push_boolean(&mut self) {
+        self.0.push(ValueType::Boolean);
+    }
+
+    fn push_any(&mut self) {
+        self.0.push(ValueType::Any);
+    }
+
+    fn push_null(&mut self) {
+        self.0.push(ValueType::Null);
+    }
+
+    fn push(&mut self, value: ValueType<'gc>) {
+        self.0.push(value);
+    }
+
+    fn pop(&mut self) -> Option<ValueType<'gc>> {
+        self.0.pop()
+    }
+
+    pub fn pop_for_multiname(&mut self, multiname: Gc<'gc, Multiname<'gc>>) {
+        if multiname.has_lazy_name() {
+            self.0.pop();
+        }
+        if multiname.has_lazy_ns() {
+            self.0.pop();
+        }
+    }
+
+    fn popn(&mut self, count: u32) {
+        for _ in 0..count {
+            self.pop();
+        }
+    }
+
+    fn clear(&mut self) {
+        self.0 = Vec::new();
```

Comment:
Let's just call `self.0.clear()`? This way we don't alloc-free the stack repeatedly.